File: 3_LongestSubstringWithoutRepeatingCharacters/3.cpp
// Time Limit Exceeded

class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int max_len = 0;
        map<char, int> char_set;

        for (int i = 0, start = 0; i < s.length(); i++){
            char c = s[i];
            map<char, int>::iterator it(char_set.find(c));

            if (it != char_set.end()){
                max_len = max(i - start, max_len);

                while(start < it->second + 1){
                    char_set.erase(char_set.find(s[start++]));
                }
            }
            char_set[c] = i;
        }

        return max_len;
    }
};

File: 3_LongestSubstringWithoutRepeatingCharacters/3.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        
        max_len = 0
        start = 0
        char_set = {}

        for i, c in enumerate(s):
            if c in char_set:
                new_start = char_set[c] + 1
                while start < new_start:
                    char_set.pop(s[start])
                    start += 1
                
            max_len = max(i - start + 1, max_len)
            char_set[c] = i

        return max_len
